otter/controller.py

```diff
@@ -150,6 +152,91 @@ def obey_config_change(log, transaction_id, config, scaling_group, state,
     return d
 
 
+def delete_group(log, trans_id, group, force):
+    """
+    Delete group based on the kind of tenant
+
+    :param log: Bound logger
+    :param str trans_id: Transaction ID of request doing this
+    :param otter.models.interface.IScalingGroup scaling_group: the scaling
+        group object
+    :param bool force: Should group be deleted even if it has servers?
+
+    :return: Deferred that fires with None
+    :raise: `GroupNotEmptyError` if group is not empty and force=False
+    """
+
+    def check_and_delete(_group, state):
+        if state.desired == 0:
+            d = trigger_convergence_deletion(log, group)
+            return d.addCallback(lambda _: state)
+        else:
+            raise GroupNotEmptyError(group.tenant_id, group.uuid)
+
+    if tenant_is_enabled(group.tenant_id, config_value):
+        return (trigger_convergence_deletion(log, group)
+                if force else group.modify_state(check_and_delete))
```

Comment:
I see. To be nitpicky, then, I have two suggestions:

1. put a comment inside `check_and_delete`, indicating that we need a lock surrounding the read and assignment of DELETING
2. put a comment in the tenant_is_enabled block indicating that we _don't_ need a lock in the force case, because we're just "atomically" updating the DELETING state
3. break the conditional-expression into a conditional-statement. This was actually a significant impediment to my understanding this code for about 10 minutes. I think a statement instead of an expression is better because it's imperative code, not just deciding between two expressions for their return values, which is where I would normally use an if-expression.